### ISSUE TITLE
make `set.add` return a boolean to indicate if insertion took place

### ISSUE THREAD

**Issue by pobrn** [python/cpython#110533]
# Feature or enhancement

### Proposal:

(I think this is a very simple idea, both conceptually and implementation-wise, that is why I did not open a thread for it on Discourse, if you'd prefer I do that, please tell me.)
(Also I have failed to find any discussions about this idea, if I had missed them, please direct me to them.)

This proposal seeks to make `set.add` return a `bool` value that indicates whether insertion took place or not. A return value of `True` would mean insertion has taken place, while a return value of `False` would mean that insertion has *not* taken place. This can simplify snippets that use `set.add` in conjunction with `set.__contains__`.

I may be wrong, but I believe this should a (mostly) backwards compatible change since `set.add` currently returns `None`, so the return value is likely not used anywhere.

For example:
```python
s = set()
...
# currently
if x in s:
    raise SomeError("duplicates not allowed")
s.add(x)

# with the proposal
if not s.add(x):
    raise SomeError("duplicates not allowed")
```

Prototype implementation (not really tested at the moment):

<details>

```diff
diff --git a/Objects/setobject.c b/Objects/setobject.c
index ae3f0b8d5e..f3f939c262 100644
--- a/Objects/setobject.c
+++ b/Objects/setobject.c
@@ -170,16 +170,16 @@ set_add_entry(PySetObject *so, PyObject *key, Py_hash_t hash)
     so->used++;
     freeslot->key = key;
     freeslot->hash = hash;
-    return 0;
+    return 1;
 
   found_unused:
     so->fill++;
     so->used++;
     entry->key = key;
     entry->hash = hash;
-    if ((size_t)so->fill*5 < mask*3)
-        return 0;
-    return set_table_resize(so, so->used>50000 ? so->used*2 : so->used*4);
+    if ((size_t)so->fill*5 >= mask*3 && set_table_resize(so, so->used>50000 ? so->used*2 : so->used*4))
+        return -1;
+    return 1;
 
   found_active:
     Py_DECREF(key);
@@ -623,7 +623,7 @@ set_merge(PySetObject *so, PyObject *otherset)
         other_entry = &other->table[i];
         key = other_entry->key;
         if (key != NULL && key != dummy) {
-            if (set_add_entry(so, key, other_entry->hash))
+            if (set_add_entry(so, key, other_entry->hash) < 0)
                 return -1;
         }
     }
@@ -904,7 +904,7 @@ set_update_internal(PySetObject *so, PyObject *other)
                 return -1;
         }
         while (_PyDict_Next(other, &pos, &key, &value, &hash)) {
-            if (set_add_entry(so, key, hash))
+            if (set_add_entry(so, key, hash) < 0)
                 return -1;
         }
         return 0;
@@ -915,7 +915,7 @@ set_update_internal(PySetObject *so, PyObject *other)
         return -1;
 
     while ((key = PyIter_Next(it)) != NULL) {
-        if (set_add_key(so, key)) {
+        if (set_add_key(so, key) < 0) {
             Py_DECREF(it);
             Py_DECREF(key);
             return -1;
@@ -1212,7 +1212,7 @@ set_intersection(PySetObject *so, PyObject *other)
                 return NULL;
             }
             if (rv) {
-                if (set_add_entry(result, key, hash)) {
+                if (set_add_entry(result, key, hash) < 0) {
                     Py_DECREF(result);
                     Py_DECREF(key);
                     return NULL;
@@ -1237,7 +1237,7 @@ set_intersection(PySetObject *so, PyObject *other)
         if (rv < 0)
             goto error;
         if (rv) {
-            if (set_add_entry(result, key, hash))
+            if (set_add_entry(result, key, hash) < 0)
                 goto error;
             if (PySet_GET_SIZE(result) >= PySet_GET_SIZE(so)) {
                 Py_DECREF(key);
@@ -1528,7 +1528,7 @@ set_difference(PySetObject *so, PyObject *other)
                 return NULL;
             }
             if (!rv) {
-                if (set_add_entry((PySetObject *)result, key, hash)) {
+                if (set_add_entry((PySetObject *)result, key, hash) < 0) {
                     Py_DECREF(result);
                     Py_DECREF(key);
                     return NULL;
@@ -1551,7 +1551,7 @@ set_difference(PySetObject *so, PyObject *other)
             return NULL;
         }
         if (!rv) {
-            if (set_add_entry((PySetObject *)result, key, hash)) {
+            if (set_add_entry((PySetObject *)result, key, hash) < 0) {
                 Py_DECREF(result);
                 Py_DECREF(key);
                 return NULL;
@@ -1631,7 +1631,7 @@ set_symmetric_difference_update(PySetObject *so, PyObject *other)
                 return NULL;
             }
             if (rv == DISCARD_NOTFOUND) {
-                if (set_add_entry(so, key, hash)) {
+                if (set_add_entry(so, key, hash) < 0) {
                     Py_DECREF(key);
                     return NULL;
                 }
@@ -1660,7 +1660,7 @@ set_symmetric_difference_update(PySetObject *so, PyObject *other)
             return NULL;
         }
         if (rv == DISCARD_NOTFOUND) {
-            if (set_add_entry(so, key, hash)) {
+            if (set_add_entry(so, key, hash) < 0) {
                 Py_DECREF(otherset);
                 Py_DECREF(key);
                 return NULL;
@@ -1834,9 +1834,11 @@ set_richcompare(PySetObject *v, PyObject *w, int op)
 static PyObject *
 set_add(PySetObject *so, PyObject *key)
 {
-    if (set_add_key(so, key))
+    int res = set_add_key(so, key);
+    if (res < 0)
         return NULL;
-    Py_RETURN_NONE;
+
+    return res ? Py_True : Py_False;
 }
 
 PyDoc_STRVAR(add_doc,
@@ -2326,7 +2328,11 @@ PySet_Add(PyObject *anyset, PyObject *key)
         PyErr_BadInternalCall();
         return -1;
     }
-    return set_add_key((PySetObject *)anyset, key);
+
+    if (set_add_key((PySetObject *)anyset, key) < 0)
+        return -1;
+
+    return 0;
 }
 
 int
```

</details>

### Has this already been discussed elsewhere?

This is a minor feature, which does not need previous discussion elsewhere

**Comment by gaogaotiantian** [comment]
I don't hate the idea of the proposal, however, from my personal perspective, this is kind of against Python's overall philosophy.

Python loves simple and explicit - so it's very rare for Python built-in functions/data structures to provide "extra info" other than what the function name implies. It either returns the value that is asked by the function, or `None` to indicate this is a `void` function.

Using your code as an example:

```python
s = set()
...
# currently
if x in s:
    raise SomeError("duplicates not allowed")
s.add(x)

# with the proposal
if not s.add(x):
    raise SomeError("duplicates not allowed")
```

It's very clear what the first piece of code tries to do, but not for the second piece - it's not obvious to the users what does it mean for `s.add(x)` to return `True`: does it mean it "successfully" added the value or it already "has" the value? I agree that this will improve performance, but I think it also sacrifices the readability.

This is just my two cents, and I might be wrong :) Even though the implementation itself is not complicated, I still believe this kind of behavior change requires a larger discussion.

**Comment by sobolevn** [comment]
Just for the reference, `PySet_Add` also does not provide such API:

```rst
.. c:function:: int PySet_Add(PyObject *set, PyObject *key)

   Add *key* to a :class:`set` instance.  Also works with :class:`frozenset`
   instances (like :c:func:`PyTuple_SetItem` it can be used to fill in the values
   of brand new frozensets before they are exposed to other code).  Return ``0`` on
   success or ``-1`` on failure. Raise a :exc:`TypeError` if the *key* is
   unhashable. Raise a :exc:`MemoryError` if there is no room to grow.  Raise a
   :exc:`SystemError` if *set* is not an instance of :class:`set` or its
   subtype.
```

**Comment by pochmann** [comment]
I've sometimes wished for that, too.

Here's a previous issue:
https://github.com/python/cpython/issues/58528

Here's a code example this would break (from a very popular [SO answer](https://stackoverflow.com/a/480227)) for removing duplicates from a sequence:
```python
def f7(seq):
    seen = set()
    seen_add = seen.add
    return [x for x in seq if not (x in seen or seen_add(x))]
```

Although if `set.add` returned a bool, that code might've never gotten written in the first place, as this would be much nicer and possibly faster:
```python
def f7(seq):
    seen = set()
    seen_add = seen.add
    return [x for x in seq if seen_add(x)]
```
Or actually:
```python
def f7(seq):
    return list(filter(set().add, seq))
```

**Comment by JelleZijlstra** [comment]
I like the idea but I think the backward compatibility problem @pochmann identifies kills it. There's bound to be code relying on `if set.add()` returning a falsy value.

**Comment by rhettinger** [comment]
We've discussed this one multiple times before (not just the ones mentioned above) and rejected the idea.  When I get a chance, I'll look for other discussion links.  IIRC, there were various issues including the ones mentioned above (backwards compatibility and not being a Pythonic pattern).  I think the killer was that in various examples we came up with, sometimes `s.add()` returning `True` look like it implied that the value was missing and was added, sometimes it looked like it implied that the value had already been present, and someone had expected it to return the element identical to the one now in the set so that is could be used for interning.  It was ambiguous, arbitrary, and not self-explanatory.
